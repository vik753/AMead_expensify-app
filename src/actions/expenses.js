import database from '../firebase/firebase';

// ADD_EXPENSE
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0,
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return database.ref('expenses')
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            }).catch((e) => console.log('Error pushing new expense: ', e));
    }
};
// REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

const startRemoveExpense = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({id}));
                // console.log('Remove succeeded!');
            })
            .catch((e) => console.log('Error: ', e.message));
    };
};
// EDIT_EXPENSE
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
// SET_EXPENSES
const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                dispatch(setExpenses(expenses));
            }).catch(e => console.log('Error loading expenses: ', e));
    };
};

export {
    startSetExpenses,
    startAddExpense,
    startRemoveExpense,
    addExpense,
    removeExpense,
    editExpense,
    setExpenses,
}