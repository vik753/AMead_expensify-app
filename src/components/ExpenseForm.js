import React from 'react';
import moment from "moment";
import {SingleDatePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

// const date = new Date();
const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : undefined,
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) :  moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (createdAt) => {
        createdAt && this.setState({createdAt: createdAt})
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description && !this.state.amount) {
            this.setState(() => ({error: 'Please provide the description and amount!'}))
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            })
        }
    };

    render() {
        return (
            <div>
                {this.state.error && (<p><b>{this.state.error}</b></p>)}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={({focused}) => this.setState({calendarFocused: focused})} // PropTypes.func.isRequired
                        numberOfMonths={1} // how months will be open in datepicker
                        firstDayOfWeek={1} // 1 - Monday
                        isOutsideRange={() => false} // lets us to choice the past dates
                        displayFormat={'DD.MM.YYYY'} // showing date format
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>{this.state.id ? 'Update expense' : 'Add Expense'}</button>
                </form>
            </div>
        );
    }
}