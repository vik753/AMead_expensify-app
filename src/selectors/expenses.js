// Get visible expenses
import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        // console.log(startDateMatch, endDateMatch);
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expenseA, expenseB) => {
        if (sortBy === 'date') {
            return expenseA.createdAt < expenseB.createdAt ? 1 : -1;
        }
        if (sortBy === 'amount') {
            return expenseA.amount < expenseB.amount ? 1 : -1;
        }
    });
};

export {getVisibleExpenses}