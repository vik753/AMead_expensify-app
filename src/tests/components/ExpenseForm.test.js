import React from 'react';
import {shallow} from 'enzyme';
import toJson from "enzyme-to-json";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

 test('should render ExpenseForm correctly', () => {
    const wrapper= shallow(<ExpenseForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
 });

test('should render ExpenseForm with data correctly', () => {
    const wrapper = shallow(<ExpenseForm  expense={expenses[0]}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
});