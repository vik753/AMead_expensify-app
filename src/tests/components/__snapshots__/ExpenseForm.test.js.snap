// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`should render ExpenseForm correctly with expense data 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`should render error for invalid form submission 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`should render error for invalid form submission 2`] = `
ShallowWrapper {
  "length": 1,
}
`;
